refactor(ivy): move directive, component and pipe factories to ngFactory

crisbeto opened the pull request:
Reworks the compiler to output the factories for directives, components and pipes under a new static field called `ngFactoryFn`, instead of the usual `factory` property in their respective defs. This should eventually allow us to inject any kind of decorated class (e.g. a pipe).

**Note:** these changes are the first part of the refactor and they don't include injectables. I decided to leave injectables for a follow-up PR, because there's some more cases we need to handle when it comes to their factories. Furthermore, directives, components and pipes make up most of the compiler output tests that need to be refactored and it'll make follow-up PRs easier to review if the tests are cleaned up now.

This is part of the larger refactor for FW-1468.

mhevery commented:
LGTM. Can @alxhub  review the changes as well, also @IgorMinar needs to approve the API changes.

trotyl commented:
Would `directiveInject()` be removed in follow-up PRs? Which could fix cross-brand-inheritance and any-brand-as-provider use cases.

petebacondarwin reviewed (approved):
For `fw-ngcc`

crisbeto committed:
refactor(ivy): move directive, component and pipe factories to ngFactoryFn

Reworks the compiler to output the factories for directives, components and pipes under a new static field called `ngFactoryFn`, instead of the usual `factory` property in their respective defs. This should eventually allow us to inject any kind of decorated class (e.g. a pipe).

**Note:** these changes are the first part of the refactor and they don't include injectables. I decided to leave injectables for a follow-up PR, because there's some more cases we need to handle when it comes to their factories. Furthermore, directives, components and pipes make up most of the compiler output tests that need to be refactored and it'll make follow-up PRs easier to review if the tests are cleaned up now.

This is part of the larger refactor for FW-1468.

crisbeto commented on packages/core/src/render3/definition.ts:
```diff
@@ -752,25 +731,35 @@ export function ɵɵdefinePipe<T>(pipeDef: {
+export function getFactoryFn<T>(type: any, throwNotFound?: boolean): FactoryFn<T>|null {
+  const factoryFn = type[NG_FACTORY_FN] || null;
+  if (!factoryFn && throwNotFound === true) {
+    throw new Error(`Type ${stringify(type)} does not have 'ngFactoryFn' property.`);
```
Done, although the same is happening below for the `ngModuleDef`. Should I do a separate PR to make it tree-shakeable as well?

crisbeto commented:
All of the feedback has been addressed @kara @alxhub. Can you take another look?

kara commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -177,7 +180,7 @@ export interface DirectiveDef<T> extends ɵɵBaseDef<T> {
   /**
    * Factory function used to create a new directive instance.
```
Can we add a quick note about why it can be null? e.g.

```suggestion
   * Factory function used to create a new directive instance.
   * Will be null initially. Populated when the factory is first requested by directive instantiation logic.
```

kara reviewed (approved):
LGTM, 1 comment nit

crisbeto committed:
fixup! refactor(ivy): move directive, component and pipe factories to ngFactoryFn

crisbeto commented:
Addressed the latest set of feedback.

AndrewKushnir commented:
@crisbeto could you please rebase and resolve the conflict (I'll run g3 presubmit after that)? Thank you.

crisbeto commented:
Rebased @AndrewKushnir.

AndrewKushnir commented:
Thanks @crisbeto. I've started presubmits (we can run them again if there are any additional changes):

- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:263806428:BASE:263806467:1565978770530:de36b073)
- [Ivy Presubmit](http://test/OCL:263806428:BASE:263806806:1565978847775:6fbd52de)

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/factory.ts:
```diff
@@ -0,0 +1,26 @@
+import {CompileResult} from '../../transform';
+
+export function getNgFactoryFnCompileResult(
+    metadata: R3FactoryFnMetadata, metadataStatement: Statement | null,
```
Let's move `metadataStatement` out of here since it's not a concern of the factory system. Also, we can put `isPipe` in `R3FactoryFnMetadata`.

alxhub commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -77,6 +82,13 @@ export interface R3ExpressionFactoryMetadata extends R3ConstructorFactoryMetadat
+  name: string;
+  type: o.Expression;
+  typeArgumentCount: number;
+  deps: R3DependencyMetadata[]|null;
```
I don't think this should be `null`able. A factory function _always_ requires dependencies (it might be `[]`, though).

alxhub replied:
You're right, actually, it should be nullable. `null` means the deps were inherited from the base class.

alxhub commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -143,8 +155,8 @@ export interface R3DependencyMetadata {
  */
-export function compileFactoryFunction(
-    meta: R3FactoryMetadata, isPipe = false): {factory: o.Expression, statements: o.Statement[]} {
+export function compileFactoryFunction(meta: R3FactoryMetadata, isPipe = false):
```
As mentioned, move `isPipe` into `R3FactoryMetadata`.

alxhub commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -234,9 +246,27 @@ export function compileFactoryFunction(
+        type: meta.type,
+        deps: meta.deps,
+        typeArgumentCount: meta.typeArgumentCount,
+        // TODO(crisbeto): this should be refactored once we start using it for injectables.
```
We can pass the injection function to use in `R3FactoryFnMetadata`.

crisbeto replied:
That's what I meant with the TODO, but when I tried doing it I had some issues when importing the proper symbols. I decided to leave it for the next PR since it's not necessary in this one and the current PR is big enough as it is.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -98,13 +81,48 @@ export function compileComponent(type: Type<any>, metadata: Component): void {
   compileInjectable(type);
 }
 
+function getComponentCompilerOutput(type: Type<any>, metadata: Component) {
```
Maybe `compileComponentFields`? Also please explicitly specify the return type.

alxhub commented on packages/core/src/render3/jit/directive.ts:
```diff
@@ -38,43 +38,26 @@ import {flushModuleScopingQueueAsMuchAsPossible, patchComponentDefWithScope, tra
+  Object.defineProperty(type, NG_FACTORY_FN, {
+    get: () => {
+      if (ngFactoryFn === null) {
+        [ngComponentDef, ngFactoryFn] = getComponentCompilerOutput(type, metadata);
```
Doesn't this result in jitting the componentdef/factorydef twice?

Reading `ngFactoryFn` should only JIT the factory, and reading `ngComponentDef` should only JIT the component.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/factory.ts:
```diff
@@ -0,0 +1,26 @@
+
+import {CompileResult} from '../../transform';
+
+export function getNgFactoryFnCompileResult(
```
```suggestion
export function compileNgFactoryField(
```

alxhub commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -129,27 +138,32 @@ export class CompilerFacadeImpl implements CompilerFacade {
+    return [
+      this.jitExpression(
+          res.expression, angularCoreEnv, jitExpressionSourceMap, constantPool.statements),
+      this.jitExpression(
```
This function (`compileComponent`) and the other decorator compilers on the facade should not be involved in factory generation. We should introduce an additional facade method `compileFactory` for this purpose.

crisbeto commented:
@alxhub I've implemented (almost) all of the comments from the latest set of feedback. Can you take another look? It's all in the last `fixup` commit. 

Note that I had to revert the changes [from this comment](https://github.com/angular/angular/pull/31953#discussion_r314834983), because making the `deps` non-nullable ended up breaking one test in ngtsc ("generates base factories for directives"). It looks like we have some logic that treats the `deps` differently depending on whether they're `null` or an empty array (see `validateConstructorDependencies`). Furthermore, I went with making it nullable in the first place, because that's how we're treating it in the metadata for pipes, directives and components, and it makes it easier to pass the metadata around.

alxhub commented on packages/compiler/src/compiler_facade_interface.ts:
```diff
@@ -39,6 +39,10 @@ export interface CompilerFacade {
       any;
+  compileFactory(
+      angularCoreEnv: CoreEnvironment, sourceMapUrl: string,
+      meta: R3PipeMetadataFacade|R3DirectiveMetadataFacade|R3ComponentMetadataFacade,
```
Technically speaking, this isn't correct. `compileFactory` should take an `R3FactoryMetadataFacade` interface, and we should remove `deps`, etc. completely from `R3DirectiveMetadata` and its facade.

Regardless, this is okay for now. Let's get this change in and clean it up later.

crisbeto committed:
fixup! refactor(ivy): move directive, component and pipe factories to ngFactoryFn

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:264402681:BASE:264402706:1566319412929:d35bb643)
- [Ivy Presubmit](http://test/OCL:264402681:BASE:264402891:1566319469675:946b1f04)

AndrewKushnir commented:
FYI, discussed with @crisbeto offline: due to the number of files changed in this PR, I will run Global TAP tonight, to make sure it's not breaking any g3 targets.

IgorMinar commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -309,6 +310,7 @@ export declare class NgStyleBase {
         [key: string]: any;
     } | null;
     static ngDirectiveDef: any;
+    static ngFactoryFn: any;
```
I believe the type of this as well as the ngDirectiveDef above should be `never` as we don't want to commit to these static properties be a public api. Can you please update both? thanks

IgorMinar commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -225,6 +225,7 @@ export declare class NgClassBase {
         [key: string]: any;
     } | null;
     static ngDirectiveDef: any;
+    static ngFactoryFn: any;
```
same as below.

IgorMinar commented on packages/core/test/render3/view_container_ref_spec.ts:
```diff
@@ -239,27 +239,28 @@ describe('ViewContainerRef', () => {
+          static ngFactoryFn = () => new AppComp(
+              ɵɵdirectiveInject(ViewContainerRef as any), injectComponentFactoryResolver())
+
+              static ngComponentDef = ɵɵdefineComponent({
```
isn't the indentation off here?

crisbeto replied:
It is. For some reason `clang-format` behaves weirdly with multiple static properties. It seems to have gotten back to normal if I turn it into a multi-line arrow function.

IgorMinar commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -309,6 +310,7 @@ export declare class NgStyleBase {
         [key: string]: any;
     } | null;
     static ngDirectiveDef: any;
+    static ngFactoryFn: any;
```
have we considered calling this field just `ngFactory` as opposed to `ngFactoryFn`? "Fn" suffix in api of this importance is a red flag to me, especially since in the case of a "Factory" it doesn't really add more semantic value to the reader.

mhevery replied:
Sorry to be even more pain, but why is it not `ngFactoryDef` to be consistent with the other naming?

AndrewKushnir commented:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:264473317:BASE:264473567:1566338047364:ac8fbf5f)
- [VE Global Presubmit](https://test.corp.google.com/OCL:264473317:BASE:264517331:1566352895169:cb228f9f)
- [Ivy Presubmit](http://test/OCL:264472911:BASE:264473913:1566338159680:740f21c5)
- [Ivy Global Presubmit](https://test.corp.google.com/OCL:264472911:BASE:264517673:1566352981362:90ee0a23)

crisbeto commented:
@IgorMinar I've addressed the feedback. Can you take another look? Note that I had to revert the change about setting the `ngDirectiveDef` and `ngFactory` of `NgClass` and `NgStyle` to `never`, because they caused some compilation errors like:

```
node_modules/@angular/common/src/directives/ng_class.d.ts:68:22 - error TS2417: Class static side 'typeof NgClass' incorrectly extends base class static side 'typeof NgClassBase'.
  Types of property 'ngDirectiveDef' are incompatible.
    Type 'DirectiveDef<NgClass>' is not assignable to type 'never'.
```

mhevery commented on packages/common/src/directives/ng_class.ts:
```diff
@@ -63,6 +66,7 @@ export const ngClassDirectiveDef = ngClassDirectiveDef__PRE_R3__;
  */
 export class NgClassBase {
   static ngDirectiveDef: any = ngClassDirectiveDef;
+  static ngFactory: any = ngClassFactory;
```
this should be `ngFactoryDef` to be consistent with `ngDirectiveDef` etc...

AndrewKushnir commented:
FYI, Global TAP runs for VE and Ivy look normal. Thank you.

crisbeto committed:
fixup! refactor(ivy): move directive, component and pipe factories to ngFactoryFn
